text/0000-pi-types.md

```diff
@@ -0,0 +1,379 @@
+- Feature Name: pi-types
+- Start Date: 2016-06-22
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+We propose a simple, yet sufficiently expressive, addition of dependent-types
+(also known as, Π-types and value-types).
+
+Type checking will not require SMT-solvers or other forms of theorem provers.
+
+## Generic value parameters
+
+A `const` type parameter acts like a generic parameter, containing a constant
+expression. Declaring a generic parameter `const a: usize`, declares a constant
+variable `a` of type `usize`.
+
+One can create implementations, structs, enums, and traits, abstracting over
+this generic value parameter.
+
+Such a parameter acts type-like in the context of types, generics, and
+polymorphism, and value-like in the context of expressions, function bodies,
+and applications.
+
+## Compile time calculations on constant parameters
+
+Since it is simply consisting of constexprs, one can apply constant functions
+(`const fn`) to the parameter, to perform compile time, type level calculations
+on the parameter. This allows for great expressiveness as `const fn` improves.
+
+## Expression `where` bounds
+
+The second construct added is the constant expression in `where` bounds. These
+contains statements about the constant parameters, which are checked at compile
+time.
+
+## Type checking
+
+Type checking is done by using a transitive rule, such that `where` bounds must
+be implied by the caller.
+
+# Motivation
+[motivation]: #motivation
+
+An often requested feature is the "type-level numerals", which enables generic
+length arrays. The current proposals are often limited to integers or even lack
+of value maps, and other critical features.
+
+There is a whole lot of other usecases as well. These allows certain often
+requested features to live in standalone libraries (e.g., bounded-integers,
+type level arithmetics, lattice types).
+
+It allows for creating powerful abstractions without type-level hackery.
+
+# Detailed design
+[design]: #detailed-design
+
+## The new value-type construct, `const`
+
+The first construct, we will introduce one allowing us to declare `ε → τ`
+constructors (value dependent types).
+
+In a sense, `const` "bridges" between types and values.
+
+Declaring a parameter `const x: T` allows use in both expression context (as a
+value of type `T`) and type context (as a type parameter).
+
+Such a parameter is declared, like type parameters, in angle brackets.
+
+The expr behavior is described as:
+
+    ValueParameterDeclaration:
+      Π ⊢ const x: T
+      ──────────────
+      Π ⊢ x: T
+
+On the type level, we use the very same semantics as the ones generic
+parameters currently follows.
+
+## `const fn`s as Π-constructors
+
+We are interested in value dependency, but at the same time, we want to avoid
+complications such as SMT-solvers and so on.
```

Comment:
> We are interested in value dependency but want to avoid having to solve "complex" constraint problems (e.g. via SMT solvers). 

Kill the so on (it doesn't add any new information).